src/pcre2.h.in

```diff
@@ -673,7 +680,9 @@ PCRE2_EXP_DECL int PCRE2_CALL_CONVENTION \
     int (*)(pcre2_substitute_callout_block *, void *), void *); \
 PCRE2_EXP_DECL int PCRE2_CALL_CONVENTION \
   pcre2_set_substitute_case_callout(pcre2_match_context *, \
-    uint32_t (*)(uint32_t, int, void *), void *); \
+    PCRE2_SIZE (*)(PCRE2_SPTR, PCRE2_SIZE, PCRE2_UCHAR *, PCRE2_SIZE, int, \
```

Comment:
should a  size of `PCRE2_ZERO_TERMIMATED` be supported here as a second parameter?

Reply:
There's no NUL. The string which is passed to the callout is not nul-terminated, and we don't require the callout to write out a NUL at the end.

Reply:
got it, but then is a substitution sting possibly larger than any other string in the library meant to be supported, or should and error be reported in that case, and if so how?

probably PCRE2_NULL might be another interesting error to report if a mechanism to do so is being added here.

Reply:
A string cannot be larger than PCRE2_SIZE; that's supposed to be big enough for the entire address space. PCRE2 supports 32-bit (2GiB) strings in a 32-bit build, and ≈infinite-size strings in a 64-bit build.

The substitute case callout API uses PCRE2_SIZE for the string sizes, so it supports exactly the same string sizes as all the other strings in PCRE2.

The callout will always be given a non-null, non-empty string by PCRE2. It can return `~(PCRE2_SIZE)0` to simply say "something failed". We can't provide more fine-grained errors for the callout to return, because we don't know what might go wrong in the application code.

I don't think we really have a reason to allow the callout to return an error code via an `int*`. The callout can save that state itself anyway.

Reply:
the problem is tht we use (the equivalent) of `(size_t)-1` for `PCRE2_ZERO_TERMINATED`, so we need to be ble to answer sensibly if that value (or equivalent) is passed as parameter 2.

returning with an error at least, sure seems like an improvement.

Reply:
Which API exactly are you worried about? The callback API is called by us, back into library client code. We'll never pass `PCRE2_ZERO_TERMINATED` for the length parameters.

Reply:
Thank you! I appreciate your help.